src/ppom-controller.ts

```diff
@@ -1014,12 +1054,12 @@ export class PPOMController extends BaseControllerV2<
       // Invoke provider and return result
       this.#provider.sendAsync(
         createPayload(method, params),
-        (error: Error, res: any) => {
+        (error, res: JsonRpcSuccess<Json>) => {
```

Comment:
@mcmire I found a typing flaw of legacy provider `send` and `sendAsync` here:

Attempting to change the error type in this line causes an error:
```suggestion
        (error: JsonRpcError, res: JsonRpcSuccess<Json>) => {
```
`... is not assignable to parameter of type (error: unknown, providerRes?: any) => void`

The original intention of typing the callback parameter in `send`/`sendAsync` as `(error: unknown, ...) => void` was probably to allow any type to be assigned to `error`. 

However, due to the contravariance of function parameters, what ends up happening is that any type narrower than `unknown` (so all types) causes the above error when passed in as the `callback` argument. 

To fix this, `send`/`sendAsync` needs to use `never` instead of `unknown` for the `error` parameter. 

```ts
sendAsync: (req: JsonRpcRequest, callback: (error: never, providerRes?: any) => void) => void;
```

Reply:
I now understand what `never` does in argument types, so thanks for that past discussion :) But I wonder if we need to use `unknown` in this case. The `error` argument here comes from `json-rpc-engine`, and while it is most likely to be a `JsonRpcError` object, it can also be anything throwable (for instance, see [here](https://github.com/MetaMask/core/blob/7b497fa71ecd785174dc6e818a945f1e415e8f00/packages/json-rpc-engine/src/JsonRpcEngine.ts#L441)). If we allow this type to be narrowed, then it seems that we would be allowing users to make an assumption about this value that isn't true. What are your thoughts?

Reply:
If I'm understanding correctly, it seems we might be talking about different things. The issue with `error: unknown` in the `sendAsync` *definition* (the assignee) is that it prohibits all error objects of all types (except `unknown`) from being passed into `sendAsync`. For example, this would give us the same "is not assignable to" type error:
```ts
        (error: Error, res: JsonRpcSuccess<Json>) => {
```

I think you might be talking about how we should type the callback *being passed into* `sendAsync` (the assignor, line 1059). There we could use a wider type than `JsonRpcError`. 

But regardless of how `error` is typed in line 1059, the `error` in the `sendAsync` definition can't be typed as `unknown`, unless the intention is to not allow any `error` object to be passed into `sendAsync`.

Reply:
Sorry, perhaps I didn't use the right wording. What I understood you saying was that it is not possible to narrow the type of `error`, and that in order to fix this problem, we need to change the type of SafeEventEmitterProvider in `@metamask/eth-json-rpc-provider` such that the callback to `sendAsync` is typed as `(error: never, providerRes?: any) => void` rather than `(error: unknown, providerRes?: any) => void`, as it is today. Is this correct?

> The issue with `error: unknown` in the `sendAsync` definition (the assignee) is that it prohibits all error objects of all types (except `unknown`) from being passed into `sendAsync`.
>
> [...] the error in the sendAsync definition can't be typed as unknown, unless the intention is to not allow any `error` object to be passed into `sendAsync`.

Hmm, I'm not sure I follow this.

If you're referring to what TypeScript does when the callback to `sendAsync` is called, which happens in `json-rpc-engine`, there are no issues there; the `error` object is allowed to be any type (otherwise you'd see a type assertion in that package when that callback is called).

If you're referring to what TypeScript does when `sendAsync` is called with a callback, the error that you get if you try to narrow the type of `error` is, I believe, correct. There aren't any guarantees we can make about this argument, so assuming that it is a `Error` or `JsonRpcError` by typing it as such would not be accurate and would lead to runtime errors.

Does that make sense?

Reply:
Yes that makes sense thanks for the explanation! I took a closer look at `json-rpc-engine` and it looks like currently `error` could be thrown as any type along the middleware chain, so it seems sound to enforce that `error` is `unknown` throughout the callback queue. 

It does seem like the callback error type in `json-rpc-engine` could potentially be narrowed to `JsonRpcEngineCallbackError` when the request is not a `JsonRpcNotification`. Consistently calling `serializeError()` on thrown error types might be enough of a fix. I'll add a comment about this in the `any` removal ticket for `json-rpc-engine`, since most of the `any`s there seem to be in catch clauses. If the fixes result in the `error` type being defined within a certain range, then we could take another look at whether we could allow the `error` type to be narrowed.